### PR TITLE
stargazer: Implement full upstream test suite in nixosTests.stargazer

### PR DESCRIPTION
**Pull request by gaykitty**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/tests/web-servers/stargazer.nix [hunk 1]
```diff
@@ -1,4 +1,47 @@
 { pkgs, lib, ... }:
+let
+  test_src = pkgs.stargazer.src;
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:3]:
```suggestion
  inherit (pkgs.stargazer) src;
```

File: nixos/tests/web-servers/stargazer.nix [hunk 2]
```diff
@@ -1,4 +1,47 @@
...
+  test_script = pkgs.stdenv.mkDerivation rec {
+    pname = "stargazer-test-script";
+    version = pkgs.stargazer.version;
+    src = test_src;
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:7]:
```suggestion
    inherit src;
```
this simplifies things a lot

File: nixos/tests/web-servers/stargazer.nix [hunk 3]
```diff
@@ -1,4 +1,47 @@
...
+    doCheck = false;
+    installPhase = ''
+      mkdir -p $out/bin
+      cp scripts/gemini-diagnostics $out/bin/test
+    '';
+  };
+  test_env = pkgs.stdenv.mkDerivation rec {
+    pname = "stargazer-test-env";
+    version = pkgs.stargazer.version;
+    src = test_src;
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:20]:
```suggestion
    inherit src;
```

File: nixos/tests/web-servers/stargazer.nix [hunk 4]
```diff
@@ -1,4 +1,47 @@
...
+    configurePhase = "true";
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:21]:
```suggestion
    dontConfigure = true;
```
but is that truly required? If the configure file is missing, the step is just skipped

File: nixos/tests/web-servers/stargazer.nix [hunk 5]
```diff
@@ -1,4 +1,47 @@
...
+    doCheck = false;
+    installPhase = ''
+      mkdir -p $out
+      cp -r * $out/
+    '';
+  };
+  scgi_server = pkgs.stdenv.mkDerivation rec {
+    pname = "stargazer-test-scgi-server";
+    version = pkgs.stargazer.version;
+    src = test_src;
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:34]:
```suggestion
    inherit src;
```

File: nixos/tests/web-servers/stargazer.nix [hunk 6]
```diff
@@ -1,4 +1,47 @@
 { pkgs, lib, ... }:
+let
+  test_src = pkgs.stargazer.src;
+  test_script = pkgs.stdenv.mkDerivation rec {
+    pname = "stargazer-test-script";
+    version = pkgs.stargazer.version;
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:6]:
```suggestion
    inherit (pkgs.stargazer) version;
```

File: nixos/tests/web-servers/stargazer.nix [hunk 7]
```diff
@@ -1,4 +1,47 @@
...
+    dontConfigure = true;
+    doCheck = false;
+    installPhase = ''
+      mkdir -p $out/bin
+      cp scripts/gemini-diagnostics $out/bin/test
+    '';
+  };
+  test_env = pkgs.stdenv.mkDerivation rec {
+    pname = "stargazer-test-env";
+    version = pkgs.stargazer.version;
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:19]:
```suggestion
    inherit (pkgs.stargazer) version;
```

File: nixos/tests/web-servers/stargazer.nix [hunk 8]
```diff
@@ -1,4 +1,47 @@
...
+    '';
+    doCheck = false;
+    installPhase = ''
+      mkdir -p $out
+      cp -r * $out/
+    '';
+  };
+  scgi_server = pkgs.stdenv.mkDerivation rec {
+    pname = "stargazer-test-scgi-server";
+    version = pkgs.stargazer.version;
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:33]:
```suggestion
    inherit (pkgs.stargazer) version;
```

File: nixos/tests/web-servers/stargazer.nix [hunk 9]
```diff
@@ -1,4 +1,47 @@
...
+    src = test_src;
+    buildInputs = with pkgs; [ python3 ];
+    configurePhase = "true";
+    buildPhase = "true";
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:37]:
```suggestion
    dontConfigure = true;
    dontBuild = true;
```

File: nixos/tests/web-servers/stargazer.nix [hunk 10]
```diff
@@ -7,25 +50,85 @@
...
+    geminiserver.wait_for_unit("scgi_server")
+    geminiserver.wait_for_open_port(1099)
     geminiserver.wait_for_unit("stargazer")
     geminiserver.wait_for_open_port(1965)
 
-    with subtest("check is serving over gemini"):
-      response = geminiserver.succeed("${pkgs.gemget}/bin/gemget --header -o - gemini://localhost:1965")
+    with subtest("stargazer test suite"):
+      response = geminiserver.succeed("sh -c 'cd ${test_env}; ${test_script}/bin/test'")
+      # response = geminiserver.succeed("sh -c 'cd ${test_env}; sleep 1; ${test_script}/bin/test --checks TLSClaims'")
```
Comment (SuperSandro2000) [inline nixos/tests/web-servers/stargazer.nix:131]:
```suggestion
```

if we want to keep it, we should remove sleep 1 because that's racy

**Comment by gaykitty** [comment]
Thx for the suggestion, all applied.

File: nixos/tests/web-servers/stargazer.nix [hunk 11]
```diff
@@ -1,4 +1,45 @@
 { pkgs, lib, ... }:
+let
+  inherit (pkgs.stargazer) src;
```
Comment (pbsds) [inline nixos/tests/web-servers/stargazer.nix:3]:
```suggestion
```

File: nixos/tests/web-servers/stargazer.nix [hunk 12]
```diff
@@ -1,4 +1,45 @@
...
+  test_script = pkgs.stdenv.mkDerivation rec {
+    pname = "stargazer-test-script";
+    inherit (pkgs.stargazer) version;
+    inherit src;
```
Comment (pbsds) [inline nixos/tests/web-servers/stargazer.nix:7]:
```suggestion
    inherit (pkgs.stargazer) version src;
```

Comment (pbsds) [reply nixos/tests/web-servers/stargazer.nix:7]: And for the packages below

**Review by pbsds — COMMENTED** [review]
Runs fine locally and on ofborg, and the log output seems sane. LGTM

Just a final nit

**Pull request merged**